### PR TITLE
Implement SignNextProposal

### PR DESCRIPTION
**Pull request by kerzhner**
Fixes https://github.com/statechannels/go-nitro/issues/394.

Test helper functions are shared between leader channel and follower channel tests and should be factored out. Leader channel tests are still under development, so this work is not done in this PR.

### CODE REVIEW

File: channel/consensus_channel/follower_channel_test.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+var alice = types.Destination(common.HexToHash("0x0a"))
+var bob = types.Destination(common.HexToHash("0x0b"))
+var aBal = uint64(200)
+var bBal = uint64(300)
+var vAmount = uint64(5)
+var existingChannel = types.Destination{1}
+var targetChannel = types.Destination{2}
+
+// TODO these helpers and the helpers in leader_channel should be shared.
+func fp() state.FixedPart {
```
Comment (andrewgordstewart) [inline channel/consensus_channel/follower_channel_test.go:23]: Note that defining `fp` pollutes the global scope of the `consensus_channel` package: you can see below that `fp` is available inside the `follower_channel.go` file. (This is also true for the vars added on L14-L20).

Screen Shot 2022-03-28 at 11 54 38 AM

Comment (andrewgordstewart) [reply channel/consensus_channel/follower_channel_test.go:23]: FYI I don't know of a good solution to this problem -- this issue is what @NiloCK has been focusing on.

File: channel/consensus_channel/consensus_channel.go [hunk 2]
```diff
@@ -87,6 +87,13 @@ type Guarantee struct {
 	right  types.Destination
 }
 
+func (g Guarantee) Equal(g2 Guarantee) bool {
```
Comment (andrewgordstewart) [inline channel/consensus_channel/consensus_channel.go:90]: 1. Should this be a private method? (See https://github.com/statechannels/go-nitro/issues/408)
2. Should this be a pointer receiver? (See https://statechannels.slack.com/archives/C02CWRSH77X/p1648318876223959)
 
```suggestion
func (g *Guarantee) equal(g2 Guarantee) bool {
```

File: channel/consensus_channel/follower_channel.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	vars := Vars{
+		TurnNum: c.current.TurnNum,
+		Outcome: c.current.Outcome.clone(),
+	}
+
+	err := vars.Add(p)
+	if err != nil {
+		return err
+	}
+	c.proposalQueue = c.proposalQueue[1:]
```
Comment (andrewgordstewart) [inline channel/consensus_channel/follower_channel.go:63]: I believe you also need to update `c.current`.

File: channel/consensus_channel/follower_channel_test.go (hunk 1 above)
Comment (andrewgordstewart) [reply channel/consensus_channel/follower_channel_test.go:23]: I am a little confused about this, actually. [This post](https://stackoverflow.com/a/59060637) suggests that `fp` is _not_ compiled when building `go-nitro`. I'd infer from that that `fp` should not be available inside the `follower_channel.go` file.


I re-did the above experiment, with a different result:
Screen Shot 2022-03-28 at 12 49 43 PM

Indeed, when I run `go test ./...` with the errant `fp` on the right, there's no problem.
```
➜  go-nitro git:(sign-next-proposal) ✗ go test ./... 
?       github.com/statechannels/go-nitro/abi   [no test files]
ok      github.com/statechannels/go-nitro/channel       (cached)
ok      github.com/statechannels/go-nitro/channel/consensus_channel     0.544s
ok      github.com/statechannels/go-nitro/channel/state (cached)
ok      github.com/statechannels/go-nitro/channel/state/outcome (cached)
?       github.com/statechannels/go-nitro/client        [no test files]
?       github.com/statechannels/go-nitro/client/engine [no test files]
ok      github.com/statechannels/go-nitro/client/engine/chainservice    (cached)
ok      github.com/statechannels/go-nitro/client/engine/messageservice  (cached)
ok      github.com/statechannels/go-nitro/client/engine/store   (cached)
ok      github.com/statechannels/go-nitro/client_test   0.499s
?       github.com/statechannels/go-nitro/crypto        [no test files]
ok      github.com/statechannels/go-nitro/protocols     (cached)
ok      github.com/statechannels/go-nitro/protocols/directdefund        (cached)
ok      github.com/statechannels/go-nitro/protocols/directfund  (cached)
ok      github.com/statechannels/go-nitro/protocols/virtualfund (cached)
ok      github.com/statechannels/go-nitro/types (cached)
```

But, `go build` fails:
```
➜  go-nitro git:(sign-next-proposal) go build ./...
# github.com/statechannels/go-nitro/channel/consensus_channel
channel/consensus_channel/follower_channel.go:59:9: undefined: fp
```

File: channel/consensus_channel/consensus_channel.go (hunk 2 above)
Comment (kerzhner) [reply channel/consensus_channel/consensus_channel.go]: The method is made private in c6a92a1. Note that the the other existing `Guarantee` method is value receiver https://github.com/statechannels/go-nitro/blob/c6a92a164efd947496b592c958c6e52109e7885f/channel/consensus_channel/consensus_channel.go#L98

File: channel/consensus_channel/consensus_channel.go [hunk 4]
```diff
@@ -204,6 +211,16 @@ func (a Add) RightDeposit() big.Int {
 	return result
 }
 
+func (a Add) Equal(a2 Add) bool {
```
[the remark this answers is not in the document]
Comment (kerzhner) [reply channel/consensus_channel/consensus_channel.go]: Made private in e42e124. To stay consistent with the other Add method, this is kept as a pointer receiver

File: channel/consensus_channel/follower_channel_test.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+	}
+
+	err = channel.SignNextProposal(proposal, testdata.Actors.Bob.PrivateKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if channel.ConsensusTurnNum() != 0 || channel.Includes(proposal.Guarantee) {
+		t.Fatal("consensus incorrectly updated")
+	}
```
[the remark this answers is not in the document]
Comment (kerzhner) [reply channel/consensus_channel/follower_channel_test.go:117]: Yep you are right!

File: channel/consensus_channel/follower_channel.go (hunk 3 above)
Comment (kerzhner) [reply channel/consensus_channel/follower_channel.go:73]: Good catch. Implemented in 43f88be

File: channel/consensus_channel/follower_channel_test.go (hunk 1 above)
Comment (kerzhner) [reply channel/consensus_channel/follower_channel_test.go:23]: This is a good point. Let's leave this to a future PR. I think it's an open question on what the best approach is.

**Pull request merged**